### PR TITLE
Bugfix/18394 batch size finder max val batches

### LINKED ISSUE #18394: BatchSizeFinder defining max validation batches for entire training loop
### Bug description

When the BatchSizeFinder callback is used, the `steps_per_trial` parameter ends up defining how many validation batches to run during the entire length of training. This is a similar issue to that observed with the LR Finder (#17412).

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug
(Adapted from @blainehoak #17412 )
```python

import time
import torch
from torch.utils.data import DataLoader, Dataset

import lightning
from lightning.pytorch import LightningModule, Trainer
from lightning.pytorch.callbacks import BatchSizeFinder


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class DummyDataModule(lightning.LightningDataModule):
    def __init__(
        self,
        length: int,
        size: int = 32,
        batch_size: int = 32,
    ):
        super().__init__()
        self.size = size
        self.length = length
        self.batch_size = batch_size

    def train_dataloader(self):
        return DataLoader(
            RandomDataset(self.size, self.length),
            batch_size=self.batch_size,
            shuffle=True,
        )

    def val_dataloader(self):
        return DataLoader(
            RandomDataset(self.size, self.length),
            batch_size=self.batch_size,
            shuffle=False,
        )


class BoringModel(LightningModule):
    def __init__(self, size=32, lr=0.1):
        super().__init__()
        self.model = torch.nn.Linear(size, 2)
        self.lr = lr

    def forward(self, x):
        return self.model(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        time.sleep(0.01)
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        time.sleep(0.5)  # Making no. steps visible in progress bar
        self.log("valid_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.model.parameters(), lr=self.lr)


def run():
    STEPS = 13  # This ends up determining the number of validation steps
    LENGTH = 10_000
    datamodule = DummyDataModule(length=LENGTH, batch_size=32)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir="/tmp/lightning_logs",
        max_epochs=10,
        enable_model_summary=False,
        callbacks=[BatchSizeFinder(steps_per_trial=STEPS, max_trials=3)],
    )
    trainer.fit(model, datamodule=datamodule)


if __name__ == "__main__":
    run()
```

### Error messages and logs

_No response_

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:
                - NVIDIA GeForce GTX 1070
        - available:         True
        - version:           11.8
* Lightning:
        - lightning:         2.0.7
        - lightning-cloud:   0.5.37
        - lightning-utilities: 0.8.0
        - pytorch-lightning: 2.0.2
        - torch:             2.0.1
        - torchmetrics:      0.11.4
        - torchvision:       0.15.2
 </details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by BoringDonut**
## What does this PR do?

Resets the CombinedLoader attribute of the eval loop to trigger a re-initialization with the correct val_limit batches.

Fixes #18394

(Adapted from #17636)

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--18854.org.readthedocs.build/en/18854/

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]
Thanks @BoringDonut, good call to carry over the same fix from the LR finder. Could you also add a changelog entry in the bugfix section?

File: tests/tests_pytorch/tuner/test_scale_batch_size.py [hunk 1]
```diff
@@ -317,7 +317,7 @@ def test_dataloader_reset_with_scale_batch_size(tmp_path, caplog, scale_method,
     assert caplog.text.count("greater or equal than the length") == int(new_batch_size == dataset_len)
 
     assert trainer.train_dataloader.batch_size == new_batch_size
-    assert trainer.val_dataloaders.batch_size == init_batch_size
+    assert trainer.val_dataloaders.batch_size == new_batch_size
```
Comment (BoringDonut) [inline tests/tests_pytorch/tuner/test_scale_batch_size.py:320]: Please note that this line is updated for the following reason:

Current `_scale_batch_size` instantiates the `_FitLoop` with `trainer.limit_val_batches = steps_per_trial`. This results in `_EvaluationLoop._max_batches` being initialized in a [nontrivial way withing the setup_data method](https://github.com/Lightning-AI/lightning/blob/1210deb5a0ff720b0ad66e38149cd28471d760ca/src/lightning/pytorch/loops/evaluation_loop.py#L194). As such we are in the following situation:
- current `_EvaluationLoop` (within the `_FitLoop`) initialized with `_max_batches` non suitable for an actual train
- there are no easy way to reset `_max_batches` without calling to `_EvaluationLoop.setup_data`  
- calling to `_EvaluationLoop.setup_data` would change batch size to the current size that was found by the `_scale_batch_size`


Imo this is pretty reasonable to change `assert trainer.val_dataloaders.batch_size` to `new_batch_size`, as this solves aforementioned issue, and there are no guarantees that `init_batch_size` were even possible to fit into the memory in the first place

**Comment by BoringDonut** [comment]
> Thanks @BoringDonut, good call to carry over the same fix from the LR finder. Could you also add a changelog entry in the bugfix section?

@awaelchli added a changelog entry ~

File: src/lightning/pytorch/CHANGELOG.md [hunk 2]
```diff
@@ -31,6 +31,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ### Fixed
 
 - Fixed an issue when replacing an existing `last.ckpt` file with a symlink ([#18793](https://github.com/Lightning-AI/lightning/pull/18793))
+- Fixed an issue when `BatchSizeFinder` `steps_per_trial` parameter ends up defining how many validation batches to run during the entire training ([#18394](https://github.com/Lightning-AI/lightning/issues/18394))
```
Comment (Borda) [inline src/lightning/pytorch/CHANGELOG.md:34]:
```suggestion


- Fixed an issue when `BatchSizeFinder` `steps_per_trial` parameter ends up defining how many validation batches to run during the entire training ([#18394](https://github.com/Lightning-AI/lightning/issues/18394))
```

**Review by Borda — APPROVED** [review]
